app/controllers/api/v1/responses_controller.rb

```diff
@@ -1,67 +1,76 @@
-  skip_before_action :verify_authenticity_token, only: [:response_flow_from_email]
+      before_action :set_response, only: %i[show update]
+      before_action :require_user!, only: [:index, :show, :create, :update]
 
-  def index
-    render json: ResponseSerializer.new(Response.all).serializable_hash
-  end
+      def index
+        render json: ResponseSerializer.new(Response.all).serializable_hash
+      end
 
-  def show
-    render json: ResponseSerializer.new(@response).serializable_hash.merge(additional_params)
-  end
+      def show
+        render json: ResponseSerializer.new(@response).serializable_hash.merge(additional_params)
+      end
 
-  def create
-    @response = current_user.responses.build(response_params)
+      def create
+        @response = current_user.responses.build(response_params)
 
-    if @response.save
-      render json: ResponseSerializer.new(@response).serializable_hash
-    else
-      render json: {error: @response.errors }, status: 422
-    end
-  end
+        if @response.save
+          render json: ResponseSerializer.new(@response).serializable_hash
+        else
+          render json: {error: @response.errors }, status: 422
+        end
+      end
 
-  def update
-    if @response.update(response_params)
-      render json: ResponseSerializer.new(@response).serializable_hash
-    else
-      render json: {error: @response.errors }, status: 422
-    end
-  end
+      def update
+        if @response.update(response_params)
+          render json: ResponseSerializer.new(@response).serializable_hash
+        else
+          render json: {error: @response.errors }, status: 422
+        end
+      end
 
-  def response_flow_from_email
-    user = User.find_by(id: params[:user_id])
-    new_response = user.responses.build(time_period_id: params[:time_period_id],
-                                        emotion_id: params[:emotion_id],
-                                        step: 'MemeSelection')
-    sign_in user
-    existed_response ||= Response.find_by(time_period_id: params[:time_period_id], user_id: user.id)
-    if existed_response.present?
-      existed_response.update(emotion_id: params[:emotion_id])
-
-      redirect_to "/responses/#{existed_response.id}"
-    else
-      new_response.save
-
-      redirect_to "/responses/#{new_response.id}"
-    end
-  end
+      def response_flow_from_email
+        set_user
+        sign_in_user
+        return response_not_working_from_emil if params[:not_working].present?
 
-  private
+        result = ResponseFlowFromEmail.new(params, @user).call
+        return redirect_to "/responses/#{result[:response].id}" if result[:success]
 
-  def set_response
-    @response = Response.find(params[:id])
-  end
+        render json: {error: result[:error] }, status: 422
+      end
 
-  def response_params
-    params.require(:response).permit(:id, :emotion_id, :time_period_id, :step)
-  end
+      private
+
+      def set_response
```

Comment:
Since the word "set" implies saving or changing, a more accurate naming for this method could be `get_response` or `retrieve_response` (also for `get_user` or `retrieve_user`). 

However when you have a method that retrieves something, a pattern our team likes to use a lot is "memoization" ([example](https://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem))